notification/twilio/voice.go

```diff
@@ -638,3 +606,100 @@ func (v *Voice) FriendlyValue(ctx context.Context, value string) (string, error)
 	}
 	return libphonenumber.Format(num, libphonenumber.INTERNATIONAL), nil
 }
+
+// processSayBody is a helper function to apply the correct verb types
+func processSayBody(resp *twiMLResponse, msgBody string, msgPauseIndex []int) {
+	// avoid a panic if the response object or msgBody isn't setup yet
+	if resp == nil || msgBody == "" {
+		return
+	}
+	// account for legacy method of a single message before pauses existed
+	if len(msgPauseIndex) < 1 {
+		resp.Say(msgBody)
+		return
+	}
+	var indexCounter int
+	var prevPauseIndex int
+	for i, pauseIndex := range msgPauseIndex {
+		if pauseIndex <= len(msgBody) && indexCounter <= len(msgBody) && pauseIndex >= 0 {
+			if tempSay := msgBody[indexCounter:pauseIndex]; tempSay != "" {
+				resp.Say(tempSay)
+			}
+
+			// only add a pause if it won't be at the end of the message
+			if i < len(msgPauseIndex) {
+				resp.Pause()
+			}
+
+			// increment the index by the diff of the current index and the previous index
+			indexCounter += pauseIndex - prevPauseIndex
+			prevPauseIndex = pauseIndex
+		}
+	}
+	// add the last say verb
+	if tempSay := msgBody[indexCounter:]; tempSay != "" {
+		resp.Say(tempSay)
+	}
+}
+
+// buildMessage is a function that will build the VoiceOptions object with the proper message contents
+func buildMessage(prefix string, msg notification.Message) (opts *VoiceOptions, err error) {
+	if prefix == "" {
+		return nil, errors.New("No prefix provided")
+	}
+
+	opts = &VoiceOptions{
+		ValidityPeriod: time.Second * 10,
+		CallbackParams: make(url.Values),
+		Params:         make(url.Values),
+	}
+
+	var message string
+	var pauseIndex []int
+	subID := -1
+	switch t := msg.(type) {
+	case notification.AlertBundle:
+		message = fmt.Sprintf("%s with alert notifications. Service '%s' has %d unacknowledged alerts.", prefix, t.ServiceName, t.Count)
+		pauseIndex = append(pauseIndex, len(fmt.Sprintf("%s with alert notifications.", prefix)))
```

Comment:
Instead of passing these indexes around everywhere, can we handle encoding/decoding at the request and use a string slice everywhere else?

Reply:
I think it makes more sense to have it directly by the message for readability and comprehension.  If we move the index injection elsewhere, it won't really be obvious what's going on (ie, ensuring that the message text, and the injection are by each other so you can easily keep them in sync).

Reply:
Keeping it separate is basically a pseudo-type, like this:
```go
type MultiString struct {
	MsgBody string
	Indexes []int
}
```

While I agree to keep things surface-level and not add unnecessary abstraction, embedding it as two separate fields in an existing struct doesn't help with readability or comprehension over `[]string`, which will track the whole dataset, but without special knowledge or the necessity to keep indexes in sync.

Tracking indexes manually for the entire lifecycle also adds noise at every layer, like tests, where the reader needs to know why it's there, what it's doing, and how to keep it correct.

The separate-indexes-encoding stuff is ONLY relevant when building or parsing the query parameters, so we can keep that complexity deliberate and localized without maintaining it everywhere else.

https://google.github.io/styleguide/go/guide#simplicity
> - Does not assume that you already know what it is doing
> - Does not assume that you can memorize all of the preceding code

https://google.github.io/styleguide/go/guide#least-mechanism
> Where there are several ways to express the same idea, prefer the one that uses the most standard tools.

Reply:
@mastercactapus Updated to pass around a type rather than multiple inputs as suggested.